### DISCUSSION TITLE
-define trim:edges=north,east,south,wast vs -trim

### DISCUSSION THREAD

**Discussion by cube1us** [ImageMagick/ImageMagick#4776] [Help] [answered]
I am working with some images using the following command, to try and control trimming - my goal is to not trim the "south" edge.   

(The command I am running is actually a command in a "tweaked" version of multicrop2 from http://www.fmwconcepts.com/imagemagick/multicrop2/index.php, runing under cygwin bash.  The script itself seems to be doing the right thing.)

$MAGICKCMD points to the magick command in ImageMagic,  at P:\Pro\ImageMagick-7.1.0-Q16-HDRI\magick

bcolor is "#a0a0a0" (so white, but not a fully bright white - a light gray.  The "foreground" color is a pretty intense purple of microfiche, captured via a specialized camera.

derotate is "-backgroundcolor $bcolor -deskew 40%"

fuzz is optimal at about 25 (under a previous release that was 20)

If I run:

$MAGICKCMD $img -crop ${wd}x${ht}+${offx}+${offy} +repage $derotate \
	-bordercolor "$bcolor" -border 2 -fuzz ${fuzzval}% +repage $tmpA3

I do not get any trimming, as expected.

If I run:

$MAGICKCMD $img -crop ${wd}x${ht}+${offx}+${offy} +repage $derotate \
	-bordercolor "$bcolor" -border 2 -trim -fuzz ${fuzzval}% +repage $tmpA3

I get full trimming, as expected (which is a problem because I have some stuff on the "south" edge which has quite a lot of near background/border color in it)

If I run:

$MAGICKCMD $img -crop ${wd}x${ht}+${offx}+${offy} +repage $derotate \
	-bordercolor "$bcolor" -border 2 -define trim:edges=north,east,south,west -fuzz ${fuzzval}% +repage $tmpA3

I get NO trimming.  (Indeed, I also get no trimming with any set of directions in there)

I had previously tried this under version 7.0.8, which also did not work.  I have also tried "$MAGICKCMD convert" with no difference.

(BTW, I have noticed that version 7.1.0 seems to be significantly slower on my Windows machine than 7.0.8, but it may only be certain operations - the script does a lot of stuff before this final command that I need to check.)

I am relatively new to using imagemagick, but not new to coding / scripting - for 40 years.

What am I missing?

JRJ

**Comment by snibgo** [comment, marked as answer]
```
-trim -fuzz ${fuzzval}% 
```
"-fuzz" is a setting that will be used in _following_ "-trim" operations. But you have already done a trim, so it is too late to do a setting, and it has no effect.
```
-define trim:edges=north,east,south,west -fuzz ${fuzzval}%
```
Likewise the "-define". But you don't do a "-trim" at all, so neither the "-define" nor the "-fuzz" have any effect.

EDIT: I should mention that IM options are executed in the order in which they appear in the command.

**Comment by cube1us** [comment]
I had not realized that the order directed the order of operations, though that does make  sense.  Also, the sources of info I looked at led me to believe that -define trim... was used IN PLACE OF -trim (and that also seemed a bit odd to me.  <shrug>)

-fuzz ${fuzzval}% -define trim:edges=north,east,west -trim

Seemed to do what I wanted.  (I may also play with the -define for the percentage of background instead.)

Thank you very much.

**Comment by snibgo** [comment]
The wording in the "-define trim" documentation is misleading. "-define <anything>" makes no changes to images. Instead, it modifies the behaviour of operations that do change images.
